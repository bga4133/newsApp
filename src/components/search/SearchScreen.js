import { useState } from "react";
import { useForm } from "../hooks/useForm";
import queryString from "query-string";
import { NewCard } from "../news/NewCard";
import { useNavigate, useLocation } from "react-router-dom";

export const SearchScreen = () => {
  const [state, setstate] = useState([]);
  // Screen de la busqueda
  const navigate = useNavigate();
  const location = useLocation();
  const { q = "" } = queryString.parse(location.search);

  // hook para obtener el value de los campos
  const [formValues, handleInputChange] = useForm({
    searchText: q
  });

  const { searchText } = formValues;

  // Hacemos la peticion para buscar la data por id
  async function getData() {
    let response = await fetch(
      `https://jsonplaceholder.typicode.com/photos?id=${q}`
    );
    let res = await response.json();
    setstate(res);
    return res;
  }

  // pasamos el estado para el filtro de noticias
  const noticesFilter = [...state];

  // funcion del submit de la busqueda
  const handleSearch = e => {
    e.preventDefault();
    getData();
    navigate(`?q=${searchText}`);
  };
  return (
    <>
      <h1 className="mt-5 mb-5">Busqueda de noticias por ID üîç</h1>

      <div className="row">
        <div className="col-5">
          <h4>
            Buscar
            <hr />
            <form onSubmit={handleSearch}>
              <input
                type="number"
                placeholder="Buscar noticias"
                className="form-control"
                name="searchText"
                autoComplete="off"
                value={searchText}
                onChange={handleInputChange}
              />
              <button
                type="submit"
                className="btn btn-outline-primary mt-3 moreInfoButton btn-lg"
              >
                Buscar ...
              </button>
            </form>
          </h4>
        </div>
        <div className="col-7">
          <h4>Resultados</h4>
          <hr />
          {q === "" ? (
            <div className="alert alert-info animate__animated animate__pulse">
              Buscar noticias
            </div>
          ) : (
            noticesFilter.length === 0 && (
              <div className="alert alert-danger animate__animated animate__fadeIn">
                No hay resultados {q}
              </div>
            )
          )}

          {noticesFilter.map(notice => (
            <NewCard key={notice.id} {...notice} />
          ))}
        </div>
      </div>
    </>
  );
};
