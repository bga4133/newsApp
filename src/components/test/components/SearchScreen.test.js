import { mount } from "enzyme";
import { MemoryRouter } from "react-router-dom";
import { SearchScreen } from "../../search/SearchScreen";

describe("Pruebas en el <SearchScreen />", () => {
  // test al componente de search screen cada test lleva su nombre y detalla el contenido del test
  test("se debe mostrar correctamente los valores por defecto", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["search"]}>
        <SearchScreen />
      </MemoryRouter>
    );

    expect(wrapper).toMatchSnapshot();
    expect(
      wrapper
        .find(".btn-outline-primary")
        .text()
        .trim()
    ).toBe("Buscar ...");
  });
  test("debe de mostrar noticia por id y el input con el valor del queryString", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["search/?q=11"]}>
        <SearchScreen />
      </MemoryRouter>
    );

    expect(wrapper.find("input").prop("value")).toBe("11");
    expect(wrapper).toMatchSnapshot();
  });

  test("debe mostrar error si no encuentra una noticia", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["search/?q=11"]}>
        <SearchScreen />
      </MemoryRouter>
    );

    expect(
      wrapper
        .find(".alert-danger")
        .text()
        .trim()
    ).toBe("No hay resultados 11");
  });
});
