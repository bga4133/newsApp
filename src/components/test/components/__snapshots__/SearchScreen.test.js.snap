// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el <SearchScreen /> debe de mostrar noticia por id y el input con el valor del queryString 1`] = `
Array [
  <h1
    className="mt-5 mb-5"
  >
    Busqueda de noticias por ID üîç
  </h1>,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Buscar
        <hr />
        <form
          onSubmit={[Function]}
        >
          <input
            autoComplete="off"
            className="form-control"
            name="searchText"
            onChange={[Function]}
            placeholder="Buscar noticias"
            type="number"
            value="11"
          />
          <button
            className="btn btn-outline-primary mt-3 moreInfoButton btn-lg"
            type="submit"
          >
            Buscar ...
          </button>
        </form>
      </h4>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="alert alert-danger animate__animated animate__fadeIn"
      >
        No hay resultados 
        11
      </div>
    </div>
  </div>,
]
`;

exports[`Pruebas en el <SearchScreen /> se debe mostrar correctamente los valores por defecto 1`] = `
Array [
  <h1
    className="mt-5 mb-5"
  >
    Busqueda de noticias por ID üîç
  </h1>,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Buscar
        <hr />
        <form
          onSubmit={[Function]}
        >
          <input
            autoComplete="off"
            className="form-control"
            name="searchText"
            onChange={[Function]}
            placeholder="Buscar noticias"
            type="number"
            value=""
          />
          <button
            className="btn btn-outline-primary mt-3 moreInfoButton btn-lg"
            type="submit"
          >
            Buscar ...
          </button>
        </form>
      </h4>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="alert alert-info animate__animated animate__pulse"
      >
        Buscar noticias
      </div>
    </div>
  </div>,
]
`;
